FlowCrypt/src/main/java/com/flowcrypt/email/core/msg/RawBlockParser.kt

```diff
@@ -65,61 +69,109 @@ object RawBlockParser {
-    return if (Part.ATTACHMENT.equals(mimePart.disposition, ignoreCase = true)) {
-      when (treatAs(part)) {
-        TreatAs.HIDDEN -> {
-          // ignore
-          emptyList()
-        }
+    return when {
+      mimePart.isAttachment() -> {
+        when (treatAs(mimePart)) {
+          TreatAs.HIDDEN -> {
+            // ignore
+            emptyList()
+          }
 
-        TreatAs.PGP_MSG -> {
-          listOf(RawBlock(RawBlockType.PGP_MSG, part.inputStream.readBytes(), isOpenPGPMimeSigned))
-        }
+          TreatAs.PGP_MSG -> {
+            listOf(
+              RawBlock(
+                RawBlockType.PGP_MSG,
+                mimePart.inputStream.readBytes(),
+                isOpenPGPMimeSigned
+              )
+            )
+          }
 
-        TreatAs.PUBLIC_KEY -> {
-          listOf(
-            RawBlock(
-              RawBlockType.PGP_PUBLIC_KEY,
-              part.inputStream.readBytes(),
-              isOpenPGPMimeSigned
+          TreatAs.PUBLIC_KEY -> {
+            listOf(
+              RawBlock(
+                RawBlockType.PGP_PUBLIC_KEY,
+                mimePart.inputStream.readBytes(),
+                isOpenPGPMimeSigned
+              )
             )
-          )
-        }
+          }
 
-        TreatAs.PRIVATE_KEY -> {
-          listOf(
+          TreatAs.PRIVATE_KEY -> {
+            listOf(
+              RawBlock(
+                RawBlockType.PGP_PRIVATE_KEY,
+                mimePart.inputStream.readBytes(),
+                isOpenPGPMimeSigned
+              )
+            )
+          }
+
+          else -> listOf(
             RawBlock(
-              RawBlockType.PGP_PRIVATE_KEY,
-              part.inputStream.readBytes(),
+              RawBlockType.ATTACHMENT,
+              mimePart.inputStream.readBytes(),
               isOpenPGPMimeSigned
             )
           )
         }
+      }
 
-        else -> listOf(
+      mimePart.isInline() && mimePart.contentID != null -> {
+        val inputStream = if (part.encoding == "base64") {
+          //We have to use FilterInputStream to replace '-' with '+' and '_' with '/' as Gmail API
+          //uses URL-safe implementation of Base64 encoding.
+          //https://datatracker.ietf.org/doc/html/rfc4648#section-5
+          val filterInputStream =
+            object : FilterInputStream((mimePart as MimeBodyPart).rawInputStream) {
+              override fun read(b: ByteArray?): Int {
+                val byteArray = ByteArray(b?.size ?: 0)
+                val count = super.read(byteArray)
+                byteArray.withIndex().forEach { (index, b) ->
+                  if (b.toInt().toChar() == '-') {
+                    byteArray[index] = '+'.code.toByte()
+                  }
+
+                  if (b.toInt().toChar() == '_') {
+                    byteArray[index] = '/'.code.toByte()
+                  }
+                }
+                b?.let { byteArray.copyInto(it) }
+                return count
```

Comment:
Maybe write like below?
```kt
val count = super.read(b)
b?.forEachIndexed { index, byte ->
  when (byte.toInt().toChar()) {
    '-' -> b[index] = '+'.code.toByte()
    '_' -> b[index] = '/'.code.toByte()
  }
}
return count
```

Reply:
I like it. I will use it with a little modification. I still need to use a temp buffer.

Reply:
any reason why we need temp buffer?

Reply:
@DenBond7 Could you answer above question?

Reply:
Ah, sorry. I didn't see this question. 

I'd like to modify only the received bytes from the original input stream. It will be clear and describe what exactly I do. Because `b` from `override fun read(b: ByteArray?)` can contain some other bytes that I don't know how they should be handled(or not).

The main point of this method - replacing received bytes (`'-' ->  '+',  '_' ->  '/'`) from the input stream and write them to the given input byte array.

Reply:
I see. Thanks for your explanation